ggshield/oauth.py

```diff
@@ -0,0 +1,274 @@
+            type=api_token_data.get("type", ""),
+        )
+        instance_config = self.config.auth_config.instances[self.instance]
+        instance_config.account = account_config
+        self.config.save()
+
+    @property
+    def redirect_uri(self) -> str:
+        return f"http://localhost:{REDIRECT_PORT}"
+
+    @property
+    def dashboard_url(self) -> str:
+        return self.config.dashboard_url
+
+    @property
+    def api_url(self) -> str:
+        return self.config.api_url
+
+
+class RequestHandlerWrapper:
+    """
+    Utilitary class to link the server and the request handler.
+    This allows to kill the server from the request processing.
+    """
+
+    oauth_client: OAuthClient
+    complete: bool
+    error_message: Optional[str] = None
+
+    def __init__(self, oauth_client: OAuthClient) -> None:
+        self.oauth_client = oauth_client
+        self.complete = False
+        self.error_message = None
+
+    @property
+    def request_handler(self) -> Type[BaseHTTPRequestHandler]:
+        class RequestHandler(BaseHTTPRequestHandler):
+            def do_GET(self_) -> None:
+                """
+                Handle the get request only.
+                If an authorization code can be extracted from the uri, attach it to the handler
+                so it can be retrieved after the request is processed, then kill the server.
+                """
+                callback_url: str = self_.path
+                parsed_url = urlparse.urlparse(callback_url)
+                if parsed_url.path == "/":
+                    try:
+                        self.oauth_client.process_callback(callback_url)
+                    except OAuthError as error:
+                        self_._end_request()
+                        self.error_message = error.message
+                    else:
+                        self_._end_request(
+                            urlparse.urljoin(
+                                self.oauth_client.dashboard_url, "authenticated"
+                            ),
+                        )
+
+                    # indicate to the serve to stop
+                    self.complete = True
+
+            def _end_request(self_, redirect_url: Optional[str] = None) -> None:
+                """
+                End the current request. If a redirect url is provided,
+                the response will be a redirection to this url.
+                If not the response will be a user error 400
+                """
+                status_code = 301 if redirect_url is not None else 400
+                self_.send_response(status_code)
+
+                if redirect_url is not None:
+                    self_.send_header("Location", redirect_url)
+                self_.end_headers()
+
+        return RequestHandler
+
+
+class OAuthError(Exception):
+    """
+    Exception raised during the authorization exchange code process
+    Its message is catched and will be raise again as a click Exception
```

Comment:
```suggestion
    Its message is caught and will be raised again as a click Exception
```

But actually maybe you can make OAuthError inherit from ClickException? Then you don't have to raise it again?

Reply:
I can't really do that because OAuthError is raised withing the request handler which wraps the exception. I have to pass the message through the RequestHandlerWrapper to re-raise the exception out of its initial scope

Reply:
Ah OK, makes sense. Can you just fix the typos, then? :)